### ISSUE TITLE
[Kubernetes] Load Balancer may enter Failed state when updating deployment

### ISSUE THREAD

**Issue by yangl900** [Azure/acs-engine#446]
## Problem
Service could lose incoming connectivity after upgrade deployment or service. While checking the service from Kubectl, it does have a public IP address, but that connection to that IP simply timeout.

## Investigation
We found out that the load balancer was configured correctly but with a 'Failed' provisioning state. This indicate the request to update load balancer (from master SPN) eventually failed. From activity logs, it appears the same, the last update to LB was when we update the deployment. Note the NRP operation is a long running operation. It appears that the controller on master node made the request to update LB, but did not follow the long running operation till completion. As a result when the operation fails, it leaves a broken LB and break all incoming traffic.

## Mitigation
The mitigation is not difficult, simply GET the LB and PUT the exact JSON back (basically retry the last update operation). Once the operation succeed, the connectivity resumed.

## Recommended Fix
the controller should follow the Azure-AsyncOperation header to track the operation, in case it fails retry the operation.

**Comment by yangl900** [comment]
@colemickens I know the code is probably in Kubernetes, but since all ACS clusters could hit this issue potentially, I figured it's good to track it here. 

Also given this is found from Livesite and has a relatively bigger impact, I recommend we prioritize this fix.

**Comment by timbrand** [comment]
Could this issue be related to the fact that I'm currently unable to create a Service of type LoadBalancer? It seems that there's no public IP generated and the service stays in pending state. Tried it twice now and waited an hour, without success.

**Comment by colemickens** [comment]
@timbrand Please file a separate issue and include logs from kube-controller-manager. And specify if this is with ACS or ACS-Engine. 

@yangl900 Can you please check and/or provide the k-c-m logs? I would suspect it's still trying to create the LB, unless there is a glaring bug where we swallow errors (I don't see any from skimming the code quickly). Does the audit log, or the k-c-m logs say why the provisioning failed?

FYI, We just use the Go SDK and it *definitely* handles long polling (otherwise creating the service would never take 3-5 minutes and none of it would really work because there's an sequencing that only works via long polling).

**Comment by yangl900** [comment]
@colemickens  I checked the kcm logs, it only said updating LB and there is one trace saying completed. I didn't find more details. Any keyword should I look at? Unfortunately the activity log also did not say why the LB enter failed state. I understand we handles long polling, but do we verify the terminal state is 'Succeeded'?

**Comment by colemickens** [comment]
The Go SDK *should* surface any non 200s as an error and the code would react accordingly and surface an error in the logs and not complete successfully. But you have an interesting point, I've not specifically seen cases where the long poll eventually fails (more accustomed to it failing in the sync portion), nor have I looked at the code that does error handling on the long poll. We've certainly had bugs around polling behavior and handling in the past.

It might be worth filing a bug on the Azure SDK For Go (or go-autorest) to do a sanity check to make sure its not swallowing any 4xx/5xxs inadvertently before it propagates to me.

I can't remember if it's output at the default verbosity level, but when it completes it should've printed the external IP? I guess maybe you sent logs via corp email, I'll take a look on Monday.

Otherwise, I'm not quite sure what else to suspect; I've had problems finding historical stuff in audit log as well. Maybe we can look at internal logging/tracing to get an clue? I'm not even quite sure how to even instrument this right now for the future without a better idea of what went awry here...

**Comment by yangl900** [comment]
well, the long polling always return 200. Usually the caller (Kubernetes, not the SDK) is responsible for checking the 'provisionState' property value, to verify if the operation succeed or not.

Also note that, in this case the external IP did not change, we updated an existing deployment, so only the node port changed (where the LB prob need change).

btw - where is the code for this azure specific driver/controller?

**Comment by colemickens** [comment]
>well, the long polling always return 200

Ah. It would be good to know the situations in which Azure LB will take a request in the sync portion but return a failure in the async portion. I've personally never seen this. 

>Also note that, in this case the external IP did not change, we updated an existing deployment, so only the node port changed (where the LB prob need change).

Oh, this is interesting. I'd be curious to know more about the exact change

>btw - where is the code for this azure specific driver/controller?

https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_loadbalancer.go#L147 That's where we update the LB specifically. To add some context, can you tell us if this has happened repeatedly, or just once? Thanks.

**Comment by colemickens** [comment]
(made some inline edits to the above post, posting this to trigger email notification to look at the current incarnation on GitHub)

**Comment by yangl900** [comment]
The "Failed" state LB happened only once, but to 2 of our clusters. I think it was something wrong in NRP as an intermediate error, since retry the exact PUT fixed the LB.

Looking at the Kubernetes code, I don't think we check the end provisioningState. Should we fix that?

For your question - no, we didn't specify the NodePort, and update deployment changed the node port. I am guessing it is because the new pod was scheduled on same node. and that's why LB needs an update.

**Comment by colemickens** [comment]
>Looking at the Kubernetes code, I don't think we check the end provisioningState. Should we fix that?

I very much think that should be fixed in `go-autorest` so that all clients benefit from this. It's very unexpected that the long polling would swallow errors (if that is indeed the case), from an SDK perspective, in my opinion.

>For your question - no, we didn't specify the NodePort, and update deployment changed the node port. I am guessing it is because the new pod was scheduled on same node. and that's why LB needs an update.

Hm, that's confusing. I've never seen that before (not to say it's impossible), but in k8s the Service and Deployment objects are totally independent. Changing the deployment should have *absolutely zero* effect on the Service object (or the LB behind it).

**Comment by colemickens** [comment]
It looks like it checks for certain status codes when polling, rather than the provisioning status: https://github.com/Azure/go-autorest/blob/master/autorest/sender.go#L157

**Comment by yangl900** [comment]
Briefly looked at the go-autorest code. I think this is the polling method should be used by the LB CreateOrUpdate() method:

https://github.com/Azure/go-autorest/blob/8a1913be6639bf0013293832bb0ccfe221287d9c/autorest/azure/async.go#L36

And it appeared been used too:

https://github.com/Azure/azure-sdk-for-go/blob/master/arm/network/loadbalancers.go#L103

now I'm a bit confused...Do you know if we have mock test for the LB failed to update case?

btw - we also do a service update (PUT, but no property change) during deployment, I should probably mention that, not sure if that caused the node port change...

**Comment by colemickens** [comment]
>Do you know if we have mock test for the LB failed to update case?

I know that we do *not* have a mock test for it. (and had I, I probably would've mocked the sync portion. I've asked the folks to generate interfaces/mock clients for testing, as it's a bit tedious to do today due to how the clients are generated.)

>we also do a service update (PUT, but no property change) during deployment, I should probably mention that, not sure if that caused the node port change...

Interestingly, I think you're hitting a bug that we just spent a lot of time tracking down. ([Issue](https://github.com/kubernetes/kubernetes/issues/43233)) ([Fix PR](https://github.com/kubernetes/kubernetes/pull/43259))

But that still doesn't answer:

1. Why did NRP hit a provisioning failure?

2. Why did this not surface up through the SDK and out of the service controller?

I had one other thought on what might be happening -- the LB port gets updated and the rules are updated with the probes being updated. But that's actually allowed anyway, and we update the LB all at one time.

I'll try to look at NRP logs on Monday.

**Comment by yangl900** [comment]
Thanks. I will send you my resource Id if that helps. Just fun facts for your question - one of the clusters got a node NIC left in 'Failed' provisioningState, and that blocked *all* further operations of the LB. I don't think anyone touched the NIC in control plane in anyway. So, maybe it's the same case that LB fall into 'Failed' state without a control plane operation.

I don't know if it's possible, but if the controller can reconcile all of the cluster resources into goal state 'Succeeded', it could be very helpful. Probably we need assume the resources esp network can fail any time.

**Comment by mfaizanse** [comment]
Is this issue resolved? @colemickens

**Comment by anhowe** [comment]
this was fixed in https://github.com/kubernetes/kubernetes/issues/43233

**Comment by Elexy** [comment]
I think I am running into a similar issue. Being pretty new to Azure: how would you PUT the LB config? 
This has given me the json of the LB:
`az network lb show -n ... --resource-group ...`

I'd like to see if that solves my problem as well or if I'd need to open a new issue.

**Comment by priyakantpatel** [comment]
Same issue, can somebody please advise me how to "GET the LB and PUT the exact JSON back"?
k8s is not getting IP and saying pending. When I see the Azure resource, I see multiple entries for static IP's.
